### ISSUE TITLE
Backend-metadata (aux) policy: fix transform/pickle/conversion loss; pytree loss stays by design

### ISSUE THREAD

**Issue by jhuggins** [TARPS-group/prob-pipe#353]
## Summary

`NumericRecord` keeps backend metadata (xarray/pandas dims, coords, index, dtype)
in a private `_aux` map so that `to_native()` can reconstruct the original backend
object from the bare `jax.Array` leaves. That aux is currently lost across several
boundaries. **Status: the structural-transform fix is decided and in progress**
(implementation underway on top of the #338 substrate). An audit of the adjacent
surfaces expanded this issue into the single aux-policy record: what survives
where, what is being fixed, and what stays lossy **by design** — so a future
implementer neither re-litigates the fixed parts nor "fixes" the design-mandated
one.

## Background: how aux works

- `NumericRecord` stores, alongside its bare-array leaves, a private `_aux` dict
  **keyed by top-level field name** → `(hooks, blob)`, where `blob` is the
  captured backend metadata. It is populated at coercion time in
  `NumericRecord._coerce_fields` (`probpipe/core/_numeric_record.py`): a field
  whose raw value is a registered backend type (see
  `probpipe/core/_array_backend.py`, `aux_for`) gets an entry, while a nested
  `NumericRecord` child is stored *by reference* and carries its own `_aux`.
- `NumericRecord.to_native()` walks one level, calling `hooks.restore(array, blob)`
  for each aux entry and recursing into nested `NumericRecord` children. It
  returns a permissive `Record`.
- Structural transforms rebuild the root through `Record._rebuild_root`
  (`probpipe/core/record.py`) from the already-extracted bare arrays and thread
  **no** aux.
- The pytree registration's static aux is the triple
  `(event_template, name, name_is_auto)` (`_record_flatten` in `record.py`);
  `_aux` is not part of it.

## The four surfaces (decision table)

| surface | behavior today (verified on `dev/named-tree-record-alignment` @ a5aab3e) | decision |
|---|---|---|
| structural transforms (`without` / `merge` / `replace` / `with_path_names`) | top-level aux lost; nested preserved only via by-reference child reuse | **fix — in progress**: thread a filtered `_aux` through the rebuild (rules below) |
| `to_numeric()` on an already-numeric record | **not idempotent**: returns a rebuilt copy and severs the nested children's by-reference aux | **fix alongside** the transform work (same construction path); also repairs the nested example below |
| pickle / `__reduce__` | aux lost at **every** level (no by-reference escape); `reloaded == original` stays `True`, so the loss is silent | **thread `_aux` through `__reduce__`** — no technical obstacle (blobs are picklable); sequence with/before #352 (see below) |
| JAX pytree round-trip (`tree_map` / `jit` / `vmap`) | aux lost | **keep as-is — design-mandated**; no change proposed (see below) |

## Corrected examples

```python
import xarray as xr
from probpipe import NumericRecord, Record

da = xr.DataArray(...)  # carries dims / coords

# Top-level backend leaf: aux LOST after a transform (the in-progress fix)
nr = Record("nr", temps=da, extra=1.0).to_numeric()
nr.to_native()["temps"]                          # -> DataArray (restored)
nr.without("extra").to_native()["temps"]         # -> bare jax.Array; dims/coords gone

# Nested backend leaf: survives construction AND transforms via by-reference reuse,
# but only under direct construction / auto-promotion:
inner = NumericRecord("grp", temps=da)
outer = NumericRecord("outer", grp=inner, extra=1.0)
outer.without("extra").to_native().at_path("grp/temps")   # -> DataArray restored

# CORRECTION to the earlier version of this issue: appending a redundant
# .to_numeric() severs the nested aux BEFORE any transform, because
# to_numeric() on an already-numeric record rebuilds rather than returning self:
outer2 = Record("outer", grp=inner, extra=1.0).to_numeric()
outer2.at_path("grp") is inner                            # -> False
outer2.to_native().at_path("grp/temps")                   # -> bare jax.Array (no edit at all)
```

The earlier nested example claimed restoration through the `.to_numeric()` form;
verbatim execution shows it does not hold — the asymmetry is real but the
redundant conversion breaks it. Hence the `to_numeric()` idempotence row above.

## Transform aux rules (the in-progress fix)

| transform | aux rule |
|---|---|
| `without` | keep entries for surviving top-level fields |
| `replace` | drop entries for replaced fields (the new value supplies its own hook or none); keep untouched |
| `merge` | union by field name |
| `with_path_names` | rename keys to match renamed top-level fields |

Implementation friction, for the record: `_rebuild_root` lives on the base
`Record` (which has no `_aux`), so this needs a `NumericRecord`-level hook plus a
construction path that accepts carried aux (`NumericRecord.__init__` already takes
an `aux` argument; the public/promotion path does not currently expose it).

## Why the pickle surface is in scope (previously unlisted)

Unlike the transform surface — where every current `to_native()` call site is the
immediate construct-then-export shape — pickle round-trips of records are already
routine and about to become more so:

- **Today:** under `prefect_task` / `prefect_flow` with a Ray/Dask task runner,
  per-row inputs and results cross process boundaries via cloudpickle
  (`tests/core/test_record_serialization.py` exists for exactly this reason). A
  distributed run therefore silently degrades `to_native()` fidelity relative to
  the same pipeline run in-process.
- **With #352 (Prefect result caching):** `persist_result=True` pickles results to
  storage, so a **cache hit would return an observably different object than a
  cold run** (bare arrays vs restored backend types) — caching stops being
  transparent. #352's acceptance criteria currently document this as a reload
  caveat; once the `__reduce__` threading lands here, that criterion flips to
  "restores".

Equality cannot catch the loss (`__eq__` compares the coerced arrays), which is
why the regression tests below must assert the restored **native type**, not `==`.

## Why the pytree half stays as-is (design-mandated)

`design/03-values-and-distributions.md` (Pytrees note) fixes the registration:
"its static aux data is the template and name only, so provenance and annotations
do not survive a JAX transform boundary; lineage instead rides on the computation
layer." Backend `_aux` is an implementation artifact of the array-backend
registry, not a design-reference concept — but the same closure applies
technically: the static aux participates in treedef equality (the `jit` cache
key), and aux blobs contain unhashable coords/indexes, so carrying them there
would break `jit`; treating coords as traced leaves instead is semantically wrong
(they are not part of the differentiable value). Note the design constraint is
scoped to the **JAX boundary only** — it does not mandate loss across transforms,
conversion, or pickle, which is why the other three rows are fixable.

## Acceptance criteria

Transforms + conversion (in progress):
- [ ] `to_native()` after `without` / `merge` / `replace` / `with_path_names`
  restores top-level backend leaves untouched by the edit, matching the
  nested-leaf behavior.
- [ ] `to_numeric()` on an already-`NumericRecord` is identity (or at minimum
  preserves nested child references and aux); the corrected nested example above
  restores after an edit.
- [ ] Regression tests covering each transform → `to_native()`, for both top-level
  and nested backend leaves, and for the `replace` / `merge` aux semantics —
  asserting restored native types, not `==`.
- [ ] Transform docstrings in `record.py` updated to state that backend aux for
  untouched leaves survives.

Pickle (sequence with/before #352):
- [ ] `__reduce__` carries `_aux` at every nesting level; pickle and cloudpickle
  round-trips restore `to_native()` for top-level and nested backend leaves.
- [ ] A test asserts the restored native type across a dump/load cycle (the `==`
  check alone passes even when aux is lost).
- [ ] #352's reload-fidelity acceptance criterion updated from "documented as
  lost" to "restores".

Unchanged by design:
- [ ] The JAX-round-trip docstrings (`_record_flatten`, the `NumericRecord`
  pytree note) remain "does not survive", citing the design constraint.
- [ ] `design/02` / `design/03` updated only if the structural-transform contract
  wording there needs to reflect aux survival (the pytree text is already
  correct).

## Sequencing

The transform/conversion work targets the #338 substrate (`_rebuild_root`, the
promotion construction path) and lands after #338. The pickle row should land
before or with #352 so result caching never exhibits hit-dependent fidelity.

**Comment by jhuggins** [comment]
Fixed on branch `dev/named-tree-record-alignment` (PR #338). Backend (`xarray` / `pandas`) metadata now round-trips across every path where it can:

- **Structural transforms** (`without` / `merge` / `replace` / `with_path_names`) carry each surviving leaf's aux onto the result, matched by leaf-object identity so it re-associates correctly across renames and nesting depth. Top-level leaves now behave like nested ones (which were already preserved by object reuse). — `53c2ce1`
- **Pickle / cloudpickle** now preserve aux: an aux-carrying `NumericRecord` pickles through its native form (`dict(self.to_native())`, whose `xarray` / `pandas` leaves pickle themselves) and re-captures aux on load. Adding `_aux` to the payload directly was not viable — the aux-hook `capture` / `restore` are closures and are not stdlib-picklable. This covers Ray transport. — `15679eb`
- **JAX pytree round-trips** still drop aux by design: the treedef static aux is the template and name only, and coords are non-hashable arrays that cannot live there without breaking `jit`.

A value transform (`map`) also drops aux, since the leaf value itself changed.

Tests: `TestBackendAuxSurvivesStructuralTransforms` (in `tests/core/test_array_backend.py`) and `TestNumericRecordAuxPickle` (in `tests/core/test_record_serialization.py`) — top-level and nested for each path, `replace` fresh-vs-carried, `merge` from both sides, source-record-unchanged, and the aux-free fast path. Full `tests/core`: 1732 passed.

Will close when PR #338 merges.
